src/value.rs

```diff
@@ -300,6 +300,76 @@ impl Value {
             n,
         })
     }
+
+    /// Prune the value down to the given type.
+    ///
+    /// The pruned type must be _smaller than or equal to_ the current type of the value.
+    /// Otherwise, this method returns `None`.
```

Comment:
In 6436f6c221c37570cf37dc9557edb373a33019a6:

Actually, there are no checks for compatibility. Probably if you give it a bad type, the method will panic.

I would suggest just saying "the output is undefined and the function is likely to panic" and changing the `Option` return to just return `Self`.

Reply:
The type checking happens as the value is split: There are sub-tasks of the form "prune value V to type T". We match over T.

If T is unit, then the result is simply unit. V is discarded.
If T is a sum type, then V is deconstructed if it is a left or right value. A new subtask is spawned. If  V is neither left or right, then the entire algorithm returns `None`.
If T is a product type, then V is deconstructed if it is a product. Two subtasks are spawned. Otherwise, the algorithm returns `None`.

I think this is somewhat of a proof that the method never panics or enters undefined behavior.

Reply:
>I think this is somewhat of a proof that the method never panics or enters undefined behavior.

Oh, I see. I missed the `?`s in the algorithm, which are actually error return paths. Ok, never mind, I'm happy.